Add CA Bundle Probe Documentation (#1570)
kyma-project/btp-manager: docs/contributor/09-10-ca-bundle-probe.md

# CA Bundle Probe

## Overview

> The CA bundle probe is a periodic background job that checks whether the TLS certificate of the SAP Service Manager token URL is trusted by the CA bundle currently mounted on the BTP Manager pod.

Reviewer: ```suggestion
The CA bundle probe is a periodic background job that checks whether the TLS certificate of the SAP Service Manager token URL is trusted by the CA bundle currently mounted on the BTP Manager Pod.
```

> It is designed for Kyma clusters where the `rt-bootstrapper` module is active. In such clusters, a custom CA bundle is injected into pods via a volume mount named `rt-bootstrapper-certs`. The probe detects this mount and uses the custom bundle as the certificate pool for TLS verification.

Reviewer: ```suggestion
It is designed for Kyma clusters where the `rt-bootstrapper` module is active. In such clusters, a custom CA bundle is injected into Pods using a volume mount named `rt-bootstrapper-certs`. The probe detects this mount and uses the custom bundle as the certificate pool for TLS verification.
```

> The probe is **disabled by default** (`ProbeInterval: 0s`). It requires a probe image to be configured via the `PROBE_IMAGE` environment variable.

Reviewer: ```suggestion
The probe is disabled by default (`ProbeInterval: 0s`). It requires a probe image to be configured using the **PROBE_IMAGE** environment variable.
```

## How It Works

> BTP Manager runs a `ProbeRunner` as a controller-runtime `Runnable`. On each interval tick, it:

Reviewer: ```suggestion
BTP Manager runs a `ProbeRunner` as a controller-runtime `Runnable`. On each interval tick, it performs the following steps:
```

1. Deletes any leftover probe Job from a previous cycle.
2. Creates a new Kubernetes Job (`btp-manager-ca-bundle-probe`) in `kyma-system`.
3. Waits for the Job to complete (up to 5 minutes).

> 4. Reads the results from annotations on the `BtpOperator` CR.

Reviewer: ```suggestion
4. Reads the results from annotations on the `BtpOperator` custom resource (CR).
```

5. Updates the `btpmanager_credential_probe_status` Prometheus metric.

> 6. If the CA bundle hash changed and TLS is healthy, restarts the `sap-btp-operator` pods.

Reviewer: ```suggestion
6. If the CA bundle hash changed and TLS is healthy, restarts the `sap-btp-operator` Pods.
```

7. Updates the `tls-probe-last-hash` annotation on the `BtpOperator` CR.

## Probe Job

> The probe Job runs a single container using the image configured via `PROBE_IMAGE`. The Job:

Reviewer: ```suggestion
The probe Job runs a single container using the image configured with **PROBE_IMAGE**. The Job performs the following steps:
```

- Runs with `RestartPolicy: Never` and `BackoffLimit: 0`.
- Runs with Istio sidecar injection disabled (`sidecar.istio.io/inject: "false"`).

> - Uses the `btp-manager-ca-bundle-probe` ServiceAccount.

Author: Not a bug. The deployed name is `btp-manager-ca-bundle-probe` because `config/default/kustomization.yaml` applies `namePrefix: btp-manager-` to all resources. The raw manifest uses `ca-bundle-probe`, but kustomize renders it as `btp-manager-ca-bundle-probe` at deploy time — which matches both `probe_runner.go` and the documentation.

- Writes results as annotations on the `BtpOperator` CR, then exits.

### Annotations Written by the Probe

| Annotation | Values | Description |
|---|---|---|
| `tls-probe-status` | `ok`, `alert`, or empty | TLS verification result |
| `tls-probe-hash` | SHA256 hex string | Hash of the CA bundle file |
| `tls-probe-updated-at` | RFC3339 timestamp | Time the probe wrote its results |

### Annotation Managed by BTP Manager

| Annotation | Description |
|---|---|
| `tls-probe-last-hash` | Hash from the previous cycle, used to detect CA bundle rotation |

## Decision Logic

BTP Manager reads the probe annotations after each Job completes and takes the following actions:

| Mount present | TLS result | Hash changed | Action |
|---|---|---|---|
| No | ok | n/a | No action (public landscape, all good) |

> | No | failed (x509) | n/a | Warning log (system CA doesn't trust tokenurl) |

Author: Fixed in aea0fc9. The decision table now clearly labels the column "BTP Manager action" and replaces the inaccurate "Warning log" rows with "No action (probe logs internally; no annotation written)" to reflect that these scenarios result in a silent-exit cycle from BTP Manager's perspective.

| Yes | ok | No | No action (TLS healthy, no rotation) |

> | Yes | ok | Yes | Restart `sap-btp-operator` pods |

Reviewer: ```suggestion
| Yes | ok | Yes | Restart `sap-btp-operator` Pods |
```

| Yes | failed (x509) | Any | Alert metric set to 1 |
| Any | failed (other) | Any | No action (probe logs internally; no annotation written) |

> Mount detection is based solely on the presence of the `rt-bootstrapper-certs` volume mount on the BTP Manager pod (via `POD_NAME` env var).

Reviewer: ```suggestion
Mount detection is based solely on the presence of the `rt-bootstrapper-certs` volume mount on the BTP Manager Pod (using **POD_NAME** environment variable).
```

## Configuration

| Parameter | Source | Default | Description |
|---|---|---|---|
| `ProbeInterval` | ConfigMap `sap-btp-manager` / CLI flag `--probe-interval` | `0s` (disabled) | How often to run the probe. Set to `0` to disable. |
| `PROBE_IMAGE` | Environment variable | — | Container image for the probe Job. Required to enable the probe. |
| `PROBE_TOKENURL_OVERRIDE` | Environment variable | — | Override the token URL used by the probe (for testing). |

> | `PROBE_FORCE_HASH` | Environment variable | — | Force a specific hash value (for testing). |

Reviewer: ```suggestion
| **ProbeInterval** | ConfigMap `sap-btp-manager` / CLI flag `--probe-interval` | `0s` (disabled) | How often to run the probe. Set to `0` to disable. |
| **PROBE_IMAGE** | Environment variable | None | Container image for the probe Job. Required to enable the probe. |
| **PROBE_TOKENURL_OVERRIDE** | Environment variable | None | Override the token URL used by the probe (for testing). |
| **PROBE_FORCE_HASH** | Environment variable | None | Force a specific hash value (for testing). |
```

## Metric

| Metric | Type | Description |
|---|---|---|

> | `btpmanager_credential_probe_status` | Gauge | `1` when alert (CA mounted but cert not trusted), `0` otherwise |

Author: Fixed in 586a76a. The metric description now explicitly states it is **not updated** on silent-exit cycles and retains its last written value.

## RBAC

> The probe Job pod uses the `btp-manager-ca-bundle-probe` ServiceAccount, which has:

Author: Same as above — `btp-manager-ca-bundle-probe` is the correct deployed name after kustomize applies `namePrefix: btp-manager-`. No change needed here.

> | `btpmanager_credential_probe_status` | Gauge | `1` when alert (CA mounted but cert not trusted). Set to `0` when the probe writes a non-alert result. **Not updated** on silent-exit cycles (no mount + TLS ok), so the gauge retains its last written value until the next cycle where the probe writes annotations. |

Reviewer: ```suggestion
| `btpmanager_credential_probe_status` | Gauge | `1` when alert (CA mounted but cert not trusted). Set to `0` when the probe writes a non-alert result. **Not updated** on silent-exit cycles (no mount + TLS ok), so the gauge retains its last written value until the next cycle, where the probe writes annotations. |
```

Reviewer: "`1` when alert " is not quite clear to me. Can you pls rephrase it?

> The probe Job pod uses the `btp-manager-ca-bundle-probe` ServiceAccount, which has:

Reviewer: ```suggestion
The probe Job Pod uses the `btp-manager-ca-bundle-probe` ServiceAccount, which has the following permissions:
```
